### PR TITLE
feat(fbdev,sdl): support display rotation

### LINKED ISSUE #5645: Software rotation for the Linux framebuffer
### Introduce the problem

The `sw_rotate` parameter from v8 doesn't exist anymore on `master` and not all framebuffer kernel drivers appear to support hardware rotation. So this is a needed feature.

### Proposal

The docs say

> In lack of hardware display rotation support `lv_draw_sw_rotate` can be used to rotate the buffer in the `flush_cb`.

However, I'm struggling with how to use this function correctly and it doesn't appear to be used in any of the other drivers yet. Most importantly, the function seems to rotate the whole buffer but `flush_cb` is called with a subarea. Can somebody advise how to implement software rotation correctly?

### PR DESCRIPTION
**Pull request by Johennes**
### Description of the feature or fix

This builds upon https://github.com/lvgl/lvgl/pull/5669 and implements software rotation in the framebuffer and SDL drivers.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

Fixes: #5645

### CODE REVIEW

**Comment by kisvegabor** [comment]
Looks good! Could you add `lv_display_rotate_area(display, area)` to replace [these](https://github.com/lvgl/lvgl/pull/5703/files#diff-6c146742287aa8dc77abe3a2022620902eda3834a3d75454e4cc0d9afba9e827R277-R314)? Same  in the SDL driver.

**Comment by Johennes** [comment]
I've added the convenience method and the missing rotations for the SDL driver. I didn't really test the SDL driver as I don't currently have a sample project that could use it set up. The code is pretty much the same as in the fbdev driver which I did test though.

**Comment by kisvegabor** [comment]
Thank you! I've tested it and the 180 degree rotation wasn't perfect, but it was easy to fix. I've also add some docs.

**Comment by Johennes** [comment]
> Thank you! I've tested it and the 180 degree rotation wasn't perfect, but it was easy to fix. I've also add some docs.

Argh, thanks for fixing that. Not sure how I missed it. :facepalm:

**Comment by Johennes** [comment]
Hey folks, is there anything else I could do to help this get reviewed?

**Comment by kisvegabor** [comment]
We are waiting for approval from 2 maintainers. cc @XuNeo @FASTSHIFT @PGNetHun

**Review by XuNeo — DISMISSED** [review]
The code looks good to me in general. Please do give it a full test.

**Comment by Johennes** [comment]
I tested the framebuffer rotation again today with [Buffybox](https://gitlab.com/postmarketOS/buffybox/) and everything appears to work great. So if the code looks good, I think this is ready land.

File: src/drivers/display/fb/lv_linux_fbdev.c
Comment (FASTSHIFT) [inline src/drivers/display/fb/lv_linux_fbdev.c]: Can we pre-allocate memory and avoid frequent temporary releases and requests for memory?

Comment (Johennes) [reply src/drivers/display/fb/lv_linux_fbdev.c]: That's a good point. I modified the code to reuse the buffer across flushes and only re-allocate it when the display size changed.

File: src/display/lv_display.c
Comment (PGNetHun) [inline src/display/lv_display.c]: I thought you would set the properties in this order (which is more understandable): `x1`, `y1`, `x2`, `y2`

**Review by PGNetHun — DISMISSED** [review]

File: src/display/lv_display.c
Comment (Johennes) [reply src/display/lv_display.c]: It's somewhat complicated because the last two assignments to `x2` and `y1` reuse the first two assignments to `x1` and `y2`. I suppose we could expand this at the cost of being slightly more verbose but I don't feel very strongly either way.

Comment (kisvegabor) [reply src/display/lv_display.c]: I think it'sokay as it is. Not the most elegant, but IMO adding verbose comments would be overkill too.

Comment (PGNetHun) [reply src/display/lv_display.c]: > It's somewhat complicated because the last two assignments to `x2` and `y1` reuse the first two assignments to `x1` and `y2`. I suppose we could expand this at the cost of being slightly more verbose but I don't feel very strongly either way.

Oh, you are totally right, I forget that order is here important. Mea culpa.
No need for comment.

File: src/display/lv_display.c
Comment (FASTSHIFT) [inline src/display/lv_display.c]: use switch case

File: src/drivers/display/fb/lv_linux_fbdev.c
Comment (FASTSHIFT) [inline src/drivers/display/fb/lv_linux_fbdev.c]: size_t rotated_buf_size;

File: src/drivers/display/fb/lv_linux_fbdev.c
Comment (FASTSHIFT) [inline src/drivers/display/fb/lv_linux_fbdev.c]: size_t buf_size = w * h * px_size;

File: src/drivers/sdl/lv_sdl_window.c
Comment (FASTSHIFT) [inline src/drivers/sdl/lv_sdl_window.c]: size_t rotated_buf_size;

File: src/drivers/display/fb/lv_linux_fbdev.c
Comment (FASTSHIFT) [inline src/drivers/display/fb/lv_linux_fbdev.c]: use switch case

File: src/drivers/sdl/lv_sdl_window.c
Comment (FASTSHIFT) [inline src/drivers/sdl/lv_sdl_window.c]: size_t buf_size = w * h * px_size;

File: src/drivers/sdl/lv_sdl_window.c
Comment (FASTSHIFT) [inline src/drivers/sdl/lv_sdl_window.c:232]: use switch case

File: src/drivers/sdl/lv_sdl_window.c
Comment (FASTSHIFT) [inline src/drivers/sdl/lv_sdl_window.c]: uint32_t fb_stride = disp->hor_res * px_size;

File: src/drivers/sdl/lv_sdl_window.c
Comment (FASTSHIFT) [inline src/drivers/sdl/lv_sdl_window.c]: lv_memzero(dsc->fb1, stride * disp->ver_res);

**Review by FASTSHIFT — CHANGES_REQUESTED** [review]

**Review by PGNetHun — APPROVED** [review]

**Review by kisvegabor — APPROVED** [review]

**Review by XuNeo — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**
